### ISSUE TITLE
[3.x] Scene tree dock crashes when using filter in large scenes

### ISSUE THREAD

**Issue by lawnjelly** [godotengine/godot#88194]
### Tested versions
3.6 dev 354404d

not present in:
3.6 beta 4
3.6 beta 3
3.6 beta 1
3.5.3
3.5.2

### System information

Linux Mint 21.1

### Issue description

Using the filter in the editor scene tree dock can cause crashes in large scenes.

Call stack:
```
1   TreeItem::get_children             tree.cpp              356  0x437fe9c 
2   SceneTreeEditor::_update_filter    scene_tree_editor.cpp 597  0x3b6528b 
3   SceneTreeEditor::_update_filter    scene_tree_editor.cpp 599  0x3b652b2 
4   SceneTreeEditor::_update_filter    scene_tree_editor.cpp 599  0x3b652b2 
5   SceneTreeEditor::_update_tree      scene_tree_editor.cpp 582  0x3b5f655 
6   SceneTreeEditor::set_filter        scene_tree_editor.cpp 994  0x3b568e3 
7   SceneTreeDock::_filter_changed     scene_tree_dock.cpp   3003 0x3b56758 
```

### Steps to reproduce

Load TPS demo.
Open `level/geometry/scenes/structure.tscn`.
Type 3 or 4 letters into the filter box at the top of the scene tree dock.

### Minimal reproduction project (MRP)

N/A

**Comment by lawnjelly** [comment]
Still bisecting, but likely due to this PR, only one in that area:
https://github.com/godotengine/godot/pull/67347

@Mickeon

**Comment by Mickeon** [comment]
heck

**Comment by AThousandShips** [comment]
I'm not sure what's causing this specific issue but there is a possible reference after free, `memdelete(p_parent)` followed by accessing it later, unsure if `keep_for_children || keep` also means `editor_selection == false`

**Comment by Mickeon** [comment]
It's worth noting that the only major difference between 3.x and 4.x is that tree items are freed instead of just hidden, so... Something along those lines may be the cause.

**Comment by Mickeon** [comment]
I'm trying to reproduce this issue on the latest 3.x **(Windows 10 debug build)** and TPS demo to no avail.
As a matter of fact I cannot even open `structure.tscn`, same goes for `structure.glb`.

```
 Class 'Material3D' or its base class cannot be instantiated.
 res://level/geometry/models/structure.glb:Resource of unrecognized type in file: Material3D.
 Failed to load resource 'res://.import/structure.glb-21fe3543c920768091702bd640720751.scn'.
 Failed loading resource: res://.import/structure.glb-21fe3543c920768091702bd640720751.scn. Make sure resources have been imported by opening the project in the editor at least once.
 Failed loading resource: res://level/geometry/models/structure.glb. Make sure resources have been imported by opening the project in the editor at least once.
 Class 'Material3D' or its base class cannot be instantiated.
 res://level/geometry/models/structure.glb:Resource of unrecognized type in file: Material3D.
 Failed to load resource 'res://.import/structure.glb-21fe3543c920768091702bd640720751.scn'.
 Failed loading resource: res://.import/structure.glb-21fe3543c920768091702bd640720751.scn. Make sure resources have been imported by opening the project in the editor at least once.
 Failed loading resource: res://level/geometry/models/structure.glb. Make sure resources have been imported by opening the project in the editor at least once.
 scene\resources\resource_format_text.cpp:157 - Couldn't load external resource: res://level/geometry/models/structure.glb
 scene\resources\packed_scene.cpp:165 - Condition "!sdata.is_valid()" is true. Returned: nullp
```

**Comment by lawnjelly** [comment]
This is waiting on #88287 to be merged, which fixes the glb bug (introduced by ORM PR).

**Comment by Mickeon** [comment]
Good to know. Honestly I still am not exactly sure on what would cause the crash here.

**Comment by akien-mga** [comment]
Fixed by #88943.
